File: 1 Algorithmic Toolbox/Week 2 Algorithmic Warm-up/week2_algorithmic_warmup/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.py
# Uses python3
import sys

def fibonacci_partial_sum_naive(from_, to):
    sum = 0

    current = 0
    next  = 1

    for i in range(to + 1):
        if i >= from_:
            sum += current

        current, next = next, current + next

    return sum % 10

def pisano(modulo):
    previous = 1
    current = 1
 
    result = 1
    while not (previous == 0 and current == 1):  # 1
        buffer = (previous + current) % modulo  # 2
        previous = current
        current = buffer
 
        result += 1
 
    return result

def fibonacci_partial_sum_fast(from_, to_):
    period = pisano(10)
    numfrom = (from_ + 1) % period
    numto = (to_ + 2) % period
    fib = [0] * (numto + 1)
    fib[0] = 0
    fib[1] = 1
    for i in range(2, numto+1):
        fib[i] = (fib[i-1] % 10 + fib[i-2] % 10) % 10
    #print(fib)
    if ((fib[numto] - fib[numfrom]) == 0):
        return 9
    else:
        return (fib[numto] - fib[numfrom]) % 10

if __name__ == '__main__':
    input = sys.stdin.read();
    from_, to = map(int, input.split())
    print(fibonacci_partial_sum_fast(from_, to))

File: 1 Algorithmic Toolbox/Week 2 Algorithmic Warm-up/week2_algorithmic_warmup/7_last_digit_of_the_sum_of_fibonacci_numbers_again/fibonacci_partial_sum.cpp
#include <iostream>
#include <vector>
using std::vector;
using namespace std;

long long get_fibonacci_partial_sum_naive(long long from, long long to) {
    long long sum = 0;

    long long current = 0;
    long long next  = 1;

    for (long long i = 0; i <= to; ++i) {
        if (i >= from) {
            sum += current;
        }

        long long new_current = next;
        next = next + current;
        current = new_current;
    }

    return sum % 10;
}

int pisano(int modulo){
    int previous = 1;
    int current = 1;
    int result = 1;
    while (not (previous == 0 && current == 1)){
    int buffer = (previous + current) % modulo;
    previous = current;
    current = buffer;
    result += 1;
    }
    return result;
}

int get_fibonacci_partial_sum_fast(long long from, long long to) {
    int numfrom = (from + 1) % pisano(10);
    int numto = (to + 2) % pisano(10);
    int fib[numto+1];
    fib[0]=0;
    fib[1]=1;
    for(int i = 2; i <= numto; i++){
        fib[i] = (fib[i-1] % 10 + fib[i-2] % 10) % 10;
    }
    if((fib[numfrom] - fib[numto]) == 0){
        return 9;
    }
    return (fib[numfrom] - fib[numto]) % 10;
}

int main() {
    long long from, to;
    std::cin >> from >> to;
    std::cout << get_fibonacci_partial_sum_fast(from, to) << '\n';
}
